rbldnsd.8
.\" rbldnsd manpage
.\"
.TH RBLDNSD 8 "January 14, 2021"
.SH NAME
rbldnsd \- DNS daemon suitable for running DNS\-based blocklists
.SH SYNOPSIS
.B rbldnsd
options
.IR zone : dataset ...

.SH DESCRIPTION
.PP
.B rbldnsd
is a small DNS\-protocol daemon which is designed to handle
queries to DNS\-based IP\-listing or NAME\-listing services.
Such services are a simple way to share/publish a list of
IP addresses or (domain) names which are "listed" for for
some reason, for example in order to be able to refuse a
service to a client which is "listed" in some blocklist.

.PP
.B rbldnsd
is not a general\-purpose nameserver.  It will answer to
A and TXT (and SOA and NS if such RRs are specified) queries,
and has limited ability to answer to some other types of queries.

.PP
.B rbldnsd
tries to handle data from two different perspectives: given
a set (or several) of "listed entries" (e.g. IP address ranges
or domain names), it builds and serves a DNS zone.  Note the
two are not the same: list of spammer's IPs is NOT a DNS zone,
but may be represented and used as such, provided that some
additional information necessary to build complete DNS zone
(e.g. NS and SOA records, maybe A records necessary for http
to work) is available.  In this context, \fBrbldnsd\fR is very
different from other general\-purpose nameservers such as BIND
or NSD: \fBrbldnsd\fR operates with \fIdatasets\fR (sets of
entries \- IP addresses or domain names, logically grouped
together), while other general\-purpose nameservers operates
with zones.  The way how \fBrbldnsd\fR operates may be somewhat
confusing to BIND experts.

.PP
For \fBrbldnsd\fR, a building block is a dataset: e.g., set of
insecure/abuseable hosts (IP addresses), set of network ranges
that belongs to various spam operations (IP ranges), domain
names that belong to spammers (RHSBL) and so on.  Usually,
different kind of information is placed into separate file,
for easy maintenance.  From a number of such datasets, \fBrbldnsd\fR
constructs a number of DNS zones as specified on command line.
A single dataset may be used for several zones, and a single
zone may be constructed from several datasets.

.PP
.B rbldnsd
will answer queries to DNS zones specified on the command
line as a set of zone specifications.
Each zone specification consists of zone basename, dataset type
and a comma\-separated list of files that forms a given dataset:
.IR   zone : type : file , file ,...

.PP
Several zones may be specified in command line, so that
.B rbldnsd
will answer queries to any of them.  Also, a single zone
may be specified several times with different datasets,
so it is possible to form a zone from a combination
of several different dataset.  The same dataset may be
reused for several zones too (and in this case, it will
be read into memory only once).

.PP
There are several dataset formats available, each is suitable
and optimized (in terms of memory, speed and ease of use)
for a specific task.  Available dataset types may be grouped
into the following categories:
.IP
lists of IP addresses.  When a query is done to a zone with such
data, query is interpreted as an IP address in a reverse form
(similar to in\-addr.arpa zone).  If the address is
found in dataset data,
.B rbldnsd
will return A and TXT records specified in data for that IP.
This is a classical IP\-based blocklist.
.IP
lists of domain names.  Similar to list of IP addresses, but
with generic domain names instead of IPs (wildcards allowed).
This type of data may be used to form a blocklist of e.g.
sender domain names.
.IP
generic list of various types of records, as an auxilary data
to form a complete nameserver.  This format is similar to bind\-style
datafiles, but very simplified.  One may specify A, TXT, NS and MX
records here.
.IP
combined set, different datasets from the list above combined
in the single (set of) source files, for easy maintenance.
.IP
acl, or Access Control List.  This is a pseudo dataset, that works
by overweriting query results based on the requestor (peer) IP address.

.SH OPTIONS

.PP
The following options may be specified:

.IP "\fB\-u\fR \fIuser\fR[:\fIgroup\fR]"
\fBrbldnsd\fR will change its userid to the specified \fIuser\fR, which
defaults to \fBrbldns\fR, and \fIgroup\fR, which by default is the
primary group of a \fIuser\fR. \fBrbldnsd\fR will refuse to run as the root
user, since this is insecure.

.IP "\fB\-r\fR \fIrootdir\fR"
\fBrbldnsd\fR will chroot to \fIrootdir\fR if specified.  Data files
should be available inside \fIrootdir\fR.

.IP "\fB\-w\fR \fIworkdir\fR"
\fBrbldnsd\fR will change its working directory to \fIworkdir\fR
(after chrooting to \fIrootdir\fR if \fB\-r\fR option is also specified).
May be used to shorten filename paths.

.IP "\fB\-b\fR \fIaddress\fR/\fIport\fR"
This option is \fIrequired\fR.  \fBrbldnsd\fR will bind to specified
\fIaddress\fR and \fIport\fR (port defaults to port 53, domain).
Either numeric IP address or a hostname may be specified, and either
port number or service name is accepted.
It is possible to specify several addresses to listen on this way, by
repeating \fB\-b\fR option.  Additionally, if there are several addresses
listed for a hostname, \fBrbldnsd\fR will listen on all of them.
Note that \fBrbldnsd\fR will work slightly faster if only one listening
address is specified.
Note the delimiter between host and port is a slash (/), not a colon,
to be able to correctly handle IPv6 addresses.

.IP \fB\-4\fR
Use IPv4 listening socket/transport, do not attempt to use IPv6
(ignored if \fBrbldnsd\fR was built without IPv6 support).

.IP \fB\-6\fR
Use IPv6 listening socket/transport, do not attempt to use IPv4
(this option will be reported as error if IPv6 support was not
compiled in).

.IP "\fB\-t\fR \fIdefttl\fR:\fIminttl\fR:\fImaxttl\fR"
Set default reply time\-to\-live (TTL) value to be \fIdefttl\fR,
and set constraints for TTL to \fIminttl\fR and \fImaxttl\fR.  Default
applies when there's no TTL defined in a given scope (in data file),
and constraints are applied when such value provided in data.
Any of the values may be omitted, including trailing colon (:) characters,
e.g. "\fB\-t\fR\ 30" set default TTL to be 30 secound, and "\fB\-t\fR\ ::120"
or "\fB\-t\fR\ ::2m" sets maximum allowed TTL to 2 minutes.
All 3 values are in time units, with optional suffix:
\fBs\fR (secounds, default), \fBm\fR (minutes), \fBh\fR (hours),
\fBd\fR (days) or \fBw\fR (weeks).  Zero \fIminttl\fR or \fRmaxttl\fR
means no corresponding constraint will be enforced.
Default \fIdefttl\fR is 35m.

.IP "\fB\-c\fR \fIcheck\fR"
Set interval between checking for zone file changes to be \fIcheck\fR,
default is 1m (one minute).  \fBrbldnsd\fR will check zone file's last
modification time every so often, and if it detects a change, zone will
be automatically reloaded.  Setting this value to 0 disables automatic
zone change detection.  This procedure may also be triggered by sending
a SIGHUP signal to \fBrbldnsd\fR (see SIGNALS section below).

.IP \fB\-e\fR
Allow non\-network addresses to be used in CIDR ranges.  Normally,
\fBrbldnsd\fR rejects addresses such as 127.2.3.4/24, where prefix
is not within the network mask derived from bit number (in this
case, correct form is 127.2.3.0/24, note the trailing zero in prefix).
This is done in order to catch possible typos in zones (\fBrbldnsd\fR
will warn about a problem and will ignore such entry).  This option
disables checking whether the CIDR prefix fits within the network mask.

.IP \fB\-q\fR
Quick and quiet start.  Normally,
.B rbldnsd
does socket initialization and zone load in foreground, writing progress
and statistic to standard output, and aborts in case of any errors.
With this flag, nothing will be printed and first zone load will be done
in background (unless \fB\-n\fR option is also given), and zone load
errors will be non\-fatal.

.IP "\fB\-p\fR \fIpidfile\fR"
Write rbldnsd's pid to the specified \fIpidfile\fR, so it will be easily
findable.  This file gets written before entering a chroot jail (if specified)
and before changing userid, so it's ok to specify e.g. /var/run/rbldnsd.pid
here.

.IP "\fB\-l\fR \fIlogfile\fR"
Specifies a file to which log all requests made.  This file is created
after entering a chroot jail and becoming a user.  Logfiles may be quite
large, esp. on busy sites (\fBrbldnsd\fR will log \fIevery\fR recognized
request if this option is specified).  This option is mainly intended for
debugging purposes.  Upon receiption of SIGHUP signal, \fBrbldnsd\fR
reopens its logfile.  If \fIlogfile\fR prefixed with a plus sign (+),
logging will not be buffered (i.e. each line will be flushed to disk);
by default, logging is buffered to reduce system load.  Specify a single
hyphen (\-) as a filename to log to standard output (filedescriptor 1),
either buffered by default, or line-buffered if specified as `+\-'
(standard output will not be "reopened" upon receiving SIGHUP signal,
but will be flushed in case logging is buffered).

.IP "\fB\-s\fR \fIstatsfile\fR"
Specifies a file where \fBrbldnsd\fR will write a line with short statistic
summary of queries made per zone, every check (\fB\-c\fR) interval.
Format of each line is:
.nf
  \fItimestamp\fR \fIzone\fR:\fIqtot\fR:\fIqok\fR:\fIqnxd\fR:\fIbin\fR:\fIbout\fR \fIzone\fR:...
.fi
where
\fItimestamp\fR is unix time (secounds since epoch),
\fIzone\fR is the name of the base zone,
\fIqtot\fR is the total number of queries received,
\fIqok\fR \- number of positive replies,
\fIqnxd\fR \- number of NXDOMAIN replies,
\fIbin\fR is the total number of bytes read from network
(excluding IP/UDP overhead and dropped packets),
\fIbout\fR is the total number of bytes written to network.
Ther are as many such tuples as there are zones, and one extra, total
typle at the end, with \fIzone\fR being "*", like:
.nf
  1234 bl1.ex:10:5:4:311:432 bl2.ex:24:13:7:248:375 *:98:35:12:820:987
.fi
Note the total values may be larger than the sum of per-zone values,
due to queries made against unlisted zones, or bad/broken packets.

\fBRbldnsd\fR will write bare timestamp to \fIstatsfile\fR when it
is starting up, shutting down or when statistic counters are being
reset after receiving SIGUSR2 signal (see below), to indicate the
points where the counters are starting back from zero.

By default, \fBrbldnsd\fR writes absolute counter values into
\fIstatsfile\fR (number of packets (bytes) since startup or
last reset).  \fIstatsfile\fR may be prefixed with plus sign (+),
in which case \fBrbldnsd\fR will write delta values, that is,
number of packets or bytes since last write, or number of
packets (bytes) per unit of time ("incremental" mode, hence
the "+" sign).

.IP \fB\-n\fR
Do not become a daemon.  Normally, \fBrbldnsd\fR will fork and go to the
background after successful initialization.  This option disables this
behaviour.

.IP \fB\-f\fR
Request \fBrbldnsd\fR to continue processing requests during data reloads.
\fBRbldnsd\fR forks a child process to handle requests while parent
reloads the data.  This ensures smooth operations, but requires
more memory, since two copies of data is keept in memory during
reload process.

.IP \fB\-d\fR
Dump all zones to stdout in BIND format and exit.  This may be suitable
to convert easily editable rbldnsd-style data into BIND zone.  \fBrbldnsd\fR
dumps all zones as one stream, so one may want to specify only one zone
with \fB\-d\fR.  Zone file will have appropriate $ORIGIN tags.  Note
that data generated may be really huge (as BIND format isn't appropriate
for this sort of things), and some entries may not be really the same
in BIND as in \fBrbldnsd\fR (e.g., IP netblocks of large size will be
represented as wildcard entries \- 10.0.0.0/8 will become *.10; excluded
entries will be represented by a CNAME to `excluded' name, so such name
should not be present in a data set).  In this mode, \fBrbldnsd\fR ignores
\fB\-r\fR (root directory) option.

.IP \fB\-v\fR
Do not show exact version information in response to version.bind CH TXT
queries (by default \fBrbldnsd\fR responds to such queries since version
0.98).  With single \fB\-v\fR, \fBrbldnsd\fR will only return "rbldnsd"
to the caller, without the version number.  Second \fB\-v\fR disables
providing any information in response to such requests, i.e. \fBrbldnsd\fR
will return REFUSE code.

.IP \fB\-C\fR
Disable automatic on\-the\-fly uncompression of data files if this feature
is compiled in (see below).

.IP \fB\-A\fR
.IP \fB\-a\fR
Controls "laziness" of \fBrbldnsd\fR when constructing replies.
With \fB\-a\fR specified, \fBrbldnsd\fR does not add AUTH section
(containing NS records) to replies unless explicitly asked for NS
records.  It is equivalent to BIND9 "minimal\-answers" configuration
setting.  While with \fB\-A\fR specified, \fBrbldnsd\fR will always
fill in AUTH section, increasing size of replies dramatically but
allowing (caching) resolver clients to react faster to changes in
nameserver lists.  Currently (as of 0.998 version), non\-lazy
(as with \fB\-A\fR) mode is the default, but it will change in
future release.

.IP "\fB\-x\fR \fIextension\fR"
Load the given \fIextension\fR file (a dynamically-linked library, usually
with ".so" suffix).  This allows to gather custom statistics or perform other
custom tasks.  See separate document for details about building and using
extensions.  This feature is not available on all platforms, and can be
disabled at compile time.

.IP "\fB\-X\fR \fIextarg\fR"
Pass the given argument, \fIextarg\fR, to the extension loaded with \fB\-x\fR.

.SH "DATASET TYPES AND FORMATS"

.PP
Dataset files are text files which are interpreted depending on
type specified in command line.  Empty lines and lines starting
with hash character (#) or semicolon (;) are ignored, except for
a special case outlined below in section titled "Special Entries".

.PP
A (comma\-separated) list of files in dataset specification
(in \fItype\fR:\fIfile\fR,\fIfile\fR,...) is interpreted as if
all files where logically combined into one single file.

.PP
When compiled with zlib support,
.B rbldnsd
is able to read gzip\-compressed data files.  So, every \fIfile\fR in
dataset specification can be compressed with \fBgzip\fR(1), and
.B rbldnsd
will read such a file decompressing it on\-the\-fly.  This feature
may be turned off by specifying \fB\-C\fR option.

.PP
.B rbldnsd
is designed to service a DNSBL, where each entry have single
A record and optional TXT record assotiated with it.
.B rbldnsd
allows to specify A value and TXT \fItemplate\fR either for each
entry individually, or to use default A value and TXT template
pair for a group of entries.  See section "Resulting A values
and TXT templates" below for a way to specify them.

.SS "Special Entries"
.PP
If a line starts with a dollar sign ($), hash character and a dollar
sign (#$), semicolon and dollar sign (;#) or colon and a dollar sign (:$),
it is interpreted in a special way, regardless of dataset type (this
is one exception where a line starting with hash character is not
ignored \- to be able to use zone files for both \fBrbldnsd\fR and
for DJB's rbldns). The following keywords, following a dollar sign,
are recognized:

.IP "\fB$SOA\fR \fIttl origindn persondn serial refresh retry expire minttl"
Specifies SOA (Start Of Authority) record for all zones using this
dataset.  Only first SOA record is interpreted.  This is the only way
to specify SOA \- by default, \fBrbldnsd\fR will not add any SOA record
into answers, and will REFUSE to answer to certain queries (notably, SOA
query to zone's base domain name).  It is recommended, but not mandatory
to specify SOA record for every zone.  If no SOA is given, negative replies
will not be cacheable by caching nameservers.  Only one, first $SOA line is
recognized in every dataset (all subsequent $SOA lines encountered in the
same dataset are silently ignored).  When constructing a zone, SOA will
be taken from \fIfirst\fR dataset where $SOA line is found, in an order as
specified in command line, subsequent $SOA lines, if any, are ignored.
This way, one may overwrite $SOA found in 3rd party data by \fIprepending\fR
small local file to the dataset in question, listing it before any other
files.
.IP
If \fIserial\fR value specified is zero, timestamp of most recent
modified file will be substituted as \fIserial\fR.
.IP
If \fIttl\fR field is zero, default ttl (\fB\-t\fR option or last \fB$TTL\fR
value, see below) will be used.
.IP
All time fields (ttl, refresh, retry, expire, minttl) may be specified
in time units.  See \fB\-t\fR option for details.

.IP "\fB$NS\fR \fIttl\fR \fInameserverdn\fR \fInameserverdn\fI..."
Specifies NS (Name Server) records for all zones using this dataset.
Only first $NS line in a dataset is recognized, all subsequent lines
are silently ignored.  When constructing a zone from several datasets,
rbldnsd uses nameservers from $NS line in only first dataset where $NS
line is given, in command-line order, just like for $SOA record.  Only
first 32 namservers are recognized.
Individual nameserver(s) may be prefixed with a minus sign (\-), which
means this single nameserver will be ignored by \fBrbldnsd\fR.  This
is useful to temporary comment out one nameserver entry without removing
it from the list.  If \fIttl\fR is zero, default ttl will be used.
The list of NS records, just like $SOA value, are taken from the \fIfirst\fR
data file in a dataset where the $NS line is found, subsequent $NS lines,
if any, are ignored.

.IP "\fB$TTL\fR \fItime-to-live\fR"
Specifies TTL (time-to-live) value for all records in current dataset.
See also \fB\-t\fR option.  \fB$TTL\fR special overrides \fB\-t\fR value
on a per-dataset basis.

.IP "\fB$TIMESTAMP\fR \fIdstamp\fR [\fIexpires\fR]"
(experimental)
Specifies the data timestamp \fIdstamp\fR when the data has been generated,
and optionally when it will expire.  The timestamps are in form
\fIyyyy\fR:\fImm\fR:\fIdd\fR[:\fIhh\fR[:\fImi\fR[:\fIss\fR]]], where
\fIyyyy\fR is the year like 2005, \fImm\fR is the month number (01..12),
\fIdd\fR is the month day number (01..31), \fIhh\fR is hour (00..23),
\fImi\fR and \fIss\fR are minutes and secounds (00.59); hours, minutes
and secounds are optional and defaults to 0; the delimiters (either
colon or dash may be used) are optional too, but are allowed for readability.
Also, single zero (0) or dash (\-) may be used as \fIdstamp\fR and/or
\fIexpires\fR, indicating the value is not given.  \fIexpires\fR may also
be specified as \fB+\fIrel\fR, where \fIrel\fR is a time specification
(probably with suffix like s, m, h, d) as an offset to \fIdstamp\fR.
.B rbldnsd
compares \fIdstamp\fR with current timestamp and refuses to load the file
if \fIdstamp\fR specifies time in the future.  And if \fIexpires\fR is
specified,
.B rbldnsd
will refuse to service requests for that data if current time is greather
than the value specified in \fIexpires\fR field.
.IP
Note that
.B rbldnsd
will check the data expiry time every time it checks for data file
updates (when receiving SIGHUP signal or every \fB\-c\fR interval).
If automatic data reload timer (\fB\-c\fR option) is disabled,
zones will not be exipired automatically.

.IP "\fB$MAXRANGE4\fR \fIrange-size\fR"
Specifies maximum size of IPv4 range allowed for IPv4\-based datasets.
If an entry covers more IP addresses than \fIrange-size\fR, it will
be ignored (and a warning will be logged).  \fIrange-size\fR may be
specified as a number of hosts, like 256, or as network prefix lenght,
like /24 (the two are the same):
.nf
  $MAXRANGE4 /24
  $MAXRANGE4 256
.fi
This constraint is active for a dataset it is specified in, and can
be owerwritten (by subsequent $MAXRANGE statement) by a smaller value,
but can not be increased.

.IP "\fB$ENT\fR \fI[true|false]\fR"
Specifies if \fBEmpty Non Terminal\fR management should be enabled or not
for this specific dataset. Values should be "true" or "false"(default).
See also \fB\-t\fR option.

This special entry is valid only when \fBrbldnsd\fR
has been compiled with the `--enable-ent` option. See the notes for more info.

.IP "\fB$\fIn\fR \fItext\fR"
(\fIn\fR is a single digit).
Specifies a \fIsubstitution variable\fR for use as $\fIn\fR placeholders
(the \fB$\fIn\fR entries are ignored in generic daaset).
See section "Resulting A values and TXT templates" below for description
and usage examples.

.IP "\fB$=\fR \fItext\fR"
Set the base template for all individual TXT records.  See section
"Resulting A values and TXT templates" below for more information.

.SS "ip4set Dataset"
.PP
A set of IP addresses or CIDR address ranges, together with A and
TXT resulting values.  IP addresses are specified one per line, by an IP
address prefix (initial octets), complete IP address, CIDR range, or IP
prefix range (two IP prefixes or complete addresses delimited by
a dash, inclusive).  Examples, to specify 127.0.0.0/24:
.nf
  127.0.0.0/24
  127.0.0
  127/24
  127\-127.0.0
  127.0.0.0\-127.0.0.255
  127.0.0.1\-255
.fi
to specify 127.16.0.0\-127.31.255.255:
.nf
  127.16.0.0\-127.31.255.255
  127.16.0\-127.31.255
  127.16\-127.31
  127.16\-31
  127.16.0.0/12
  127.16.0/12
  127.16/12
.fi
Note that in prefix range, last boundary is completed with all\-ones (255),
not all\-zeros line with first boundary and a prefix alone.  In prefix
ranges, if last boundary is only one octet (127.16\-31), it is treated
as "suffix", as value of last \fIspecified\fR octet of the first boundary
prefix (127.16.0\-31 is treated as 127.16.0.0\-127.16.31.255, i.e.
127.16.0.0/19).
.PP
After an IP address range, A and TXT values for a given entry may be specified.
If none given, default values in current scope (see below) applies.  If a value
starts with a colon, it is interpreted as a pair of A record and TXT template,
delimited by colon (:127.0.0.2:This entry is listed).  If a value does not start
with a colon, it is interpreted as TXT template only, with A record defaulting
to the default A value in current scope.
.PP
IP address range may be followed by a comment char (either hash character (#)
or semicolon (;)), e.g.:
.nf
  127/8 ; loopback network
.fi
In this case all characters up to the end of line are ignored, and default A
and TXT values will be used for this IP range.
.PP
Every IP address that fits within any of specified ranges is "listed", and
.B rbldnsd
will respond to reverse queries against it within specified zone with
positive results.  In contrast, if an entry starts with an exclamation
sign (!), this is an
.I exclusion
entry, i.e. corresponding address range is excluded from being listed
(and any value for this record is ignored).  This may be used to specify
large range except some individual addresses, in a compact form.
.PP
If a line starts with a colon (:), this line specifies the default A value
and TXT template to return (see below) for all subsequent entries up
to end of current file.  If no default entry specified, and no value
specified for a given record, \fBrbldnsd\fR will return 127.0.0.2 for
matching A queries and no record for matching TXT queries.  If TXT
record template is specified and contains occurences of of dollar sign
($), every such occurence is replaced with an IP address in question,
so singe TXT template may be used to e.g. refer to a webpage for an
additional information for a specific IP address.

.SS "ip4trie Dataset"
.PP
Set of IP4 CIDR ranges with corresponding (A, TXT) values.  This
dataset is similar to ip4set, but uses a different internal
representation.  It accepts CIDR ranges only (not a.b.c.d\-e.f.g.h),
and allows for the specification of A/TXT values on a per CIDR range
basis.  (If multiple CIDR ranges match a query, the value for longest
matching prefix is returned.)  Exclusions are supported too.
.PP

NEWS
This file describes user-visible changes in rbldnsd.
Newer news is at the top.

1.0pre (Still not official, to be released)
 - Empty Non Terminals patch. This is a compile-time option and
   is meant to address some incompatibilities with RFC 7816.
   Adding the "$ENT" special entity to all the datasets.
 - Support for systemd added (thanks to Marco d'Itri). Please
   use --enable-systemd when configuring and make sure that you have
   pkg-config installed.
 - Test suite migrated over to python-3. Thanks to
   Michael Orlitzky for the contributions.
 - Fix a longstanding bug in the base-template replacement
 - Removal of deprecated features (aka: NS record compatibility mode)
 - Adding -F flag, used to identify the log facility of the daemon.
 - fix tests for systems without ipv6 support, or when ipv6 is
   disabled in rbldnsd at compile-time
 - fix tests for API change in pydns >= 2.3.6
 - It is no longer an error to request binding to a particular
   address/port more than once. (The subsequent requests are simply
   ignored.) (This avoids confusion on certain systems/configurations
   where gethostbyname("localhost") can return 127.0.0.1 multiple
   times.)
 - Moving rbldnsd.spec file under contrib/rpm/ to match debian move
 - Updating various pointers to point to new contrib/ locations
   (both debian and rpm spec)
 - Maintenance of RPM spec file to match new file locations, versions,
   and the make dist tar file naming schema.
 - Fix for sockaddr_equal() which was missing an IPv6 case.

0.998b (21 Dec 2016)
 - Fix for memory errors on very large datasets.
   Patch by Andrew Clayton

0.998a (06 Sep 2016)
 - Minor fixes in copyright and documentation
 - bugfix: minor fix to prevent errors on newer compilers
 - Moving debian/ old directory under contrib/debian/

0.998 (05 Dec 2015)

 - bugfix: correctly handle V4MAPPED (v4 in v6) addresses, the
   original v6 prefix was wrong.  Thanks to Alex Lasoriti for
   finding the issue
 - bugfix: sometimes IP4-based datasets gave false positives when
   an IP6 dataset were present, and it was also possible to have
   false positive in IP6 datasets.  Both has been fixed.

0.997a (23 Jul 2013)

 - minor fixes/changes in packaging, no code changes.
   In particular, fixes a build failure on *BSD introduced in 0.997

0.997 (29 Jun 2013)

 - main feature of this version is ipv6 support.  Many thanks to
   Geoffrey T. Dairiki for the implementation of btrie (btrie.c) which
   is far superior to old ip4trie code and handles both v4 and v6

 - feature: ip6trie - new dataset supports listing of arbitrary length
   ip6 CIDRs, along with individual A/TXT values for each prefix

 - feature: ip6tset - new dataset supports listing of ip6 /64 subnets
   and the exclusion of /128 subnets; only supports a single A/TXT value
   for the entire dataset

 - optimization: ip4trie - using new trie implementation (developed for
   the ip6trie dataset) decreases memory consumption by roughly a factor
   of three

 - feature: acl dataset - ip6 addresses are now supported in ACLs

 - feature: added --enable-asserts configure option to enable
   compilation of debugging assertions; assertion checking is
   disabled by default

 - featurette: zero-length "wildcard" IP4 CIDR prefixes are now
   allowed in ip4trie and acl datasets.

0.996b (29 Mar 2008)

 - most stuff is very minor, while preparing for larger changes
   for 0.997 and 1.0.

 - several (mostly minor) bug and portability fixes.

 - feature: ability to specify "base template" for a dataset, using $=
   variable.  See manpage for details, "Resulting A values and TXT templates"
   section.

 - incompat: due to the above change ($= base template), in TXT entries which
   starts with equal sign (=), the first character (which is `=', obviously)
   is removed.  $= is now treated specially too.

 - feature: (experimental) support for dynamically-loaded extensions (DSOs),
   in place of (previously unfinished) compile-time "hooks" support.
   New options -x/-X, to load an extension and to specify argument for it.
   Usage document has to be written still.

0.996a (27 Jul 2006) (The "34-Birthday Release")

 - -a/-A (-A is new, currently a noop) option clarification/addition.
   Don't mark -a as experimental anymore, and note that -a mode
   ("lazy", minimal-answers, w/o AUTH section in every reply) will
   be the default in future versions.  -A tells rbldnsd to go to
   "non-lazy" mode.  Document options in the manpage.

 - bugfix: fix configure script breakages:
   portability: for f; do => for f in "$@"; do
   fix broken GNU C (mis)detection

 - bugfix: fix dataset "un-expiration". Previously, once a dataset has expired,
   it never "un-expires" again even if new expire time is in future.  Due to
   missing reset of a dataset structure field.

 - portability: apparently at least one (broken) linux distribution includes
   kernel modification which leads to losing SIGALRM interrupts at times.
   So use setitimer() instead of alarm(), if available.

 - minor code cleanups here and there (fixing (real) GCC-4.1 warnings).

 - Debian-specific: adopt to more recent Debian packaging requiriments,
   and ensure that Debian package builds with zlib support

0.996 (19 Feb 2006)

 - experimental feature: data expiration support, in form
    $TIMESTAMP created expires
   see manpage for details.

 - feature: recognize new 'pass' entry in ACL "dataset", to allow
   whitelisting of a particular network/host covered by another
   ACL entry.

 - bugfix (sort of): deal with possible null-pointer dereferences on some
   systems such as FreeBSD where realloc(smaller_amount) may actually
   return NULL.  Note that this particular malloc implementation (where
   realloc() may return NULL if requested to reallocate to a smaller
   amount of memory) perform very badly with rbldnsd in the first place:
   rbldnsd tries to free some unused memory at the end of data load
   process, and realloc() forces a copy so there will be extra copy of

rbldnsd.py
""" Run an rbldnsd and send it DNS queries.


"""
import errno
from itertools import count
import subprocess
from tempfile import NamedTemporaryFile, TemporaryFile
import time
import unittest

try:
    import DNS
except ImportError:
    raise RuntimeError("The python3-dns library is not installed...")
try:
    from DNS import SocketError as DNS_SocketError
except ImportError:
    class DNS_SocketError(Exception):
        """ Dummy, never raised.
        """

DUMMY_ZONE_HEADER = """
$SOA 0 example.org. hostmaster.example.com. 0 1h 1h 2d 1h
$NS 1d ns0.example.org
"""

class ZoneFile(object):
    def __init__(self, lines=None, no_header=False):
        self._file = NamedTemporaryFile(delete=False)
        if not no_header:
            self._file.write(bytes(DUMMY_ZONE_HEADER, encoding = 'utf-8'))
        if lines is not None:
            self.writelines(lines)
        self._file.flush()

    def __del__(self):
        self._file.close()

    @property
    def name(self):
        return self._file.name

    def write(self, str):
        self._file.write(bytes(str))
        self._file.flush()

    def writelines(self, lines):
        self._file.writelines(bytes("%s\n" % line, encoding = 'utf-8') for line in lines)
        self._file.flush()

class DaemonError(Exception):
    """ Various errors having to do with the execution of the daemon.
    """

class QueryRefused(Exception):
    """ Query to rbldnsd was REFUSED.
    """


class Rbldnsd(object):
    def __init__(self, datasets=None,
                 daemon_addr='localhost', daemon_port=5300,
                 daemon_bin='./rbldnsd',
                 stderr=None):
        self._daemon = None
        self.datasets = []
        self.daemon_addr = daemon_addr
        self.daemon_port = daemon_port
        self.daemon_bin = daemon_bin
        self.stderr = stderr

    def add_dataset(self, ds_type, file, soa='example.com'):
        self.datasets.append((soa, ds_type, file))

    def __enter__(self):
        self._start_daemon()
        return self

    def __exit__(self, exc_type, exc_value, exc_tb):
        self._stop_daemon()

    def __del__(self):
        if self._daemon:
            self._stop_daemon()

    def query(self, name, qtype='TXT'):
        if not self._daemon:
            raise DaemonError("daemon not running")
        elif self._daemon.poll() is not None:
            raise DaemonError("daemon has died with code %d"
                              % self._daemon.returncode)

        req = DNS.Request(name=name, qtype=qtype, rd=0)
        resp = req.req(server=self.daemon_addr, port=self.daemon_port)
        status = resp.header['status']
        if status == 'REFUSED':
            raise QueryRefused("REFUSED")
        elif status == 'NXDOMAIN':
            return None
        else:
            assert status == 'NOERROR'
            assert len(resp.answers) == 1
            assert len(resp.answers[0]['data']) == 1
            return resp.answers[0]['data'][0]

    def _start_daemon(self):
        if len(self.datasets) == 0:
            raise ValueError("no datasets defined")

        cmd = [ self.daemon_bin, '-n',
                '-b', '%s/%u' % (self.daemon_addr, self.daemon_port),
                ]
        for zone, ds_type, file in self.datasets:
            if isinstance(file, str):
                filename = file
            else:
                filename = file.name
            cmd.append("%s:%s:%s" % (zone, ds_type, filename))

        self._stdout = TemporaryFile()
        self._daemon = daemon = subprocess.Popen(cmd, stdout=self._stdout,
                                                 stderr=self.stderr)

        # wait for rbldnsd to start responding
        for retry in count():
            if daemon.poll() is not None:
                raise DaemonError(
                    "rbldsnd exited unexpectedly with return code %d"
                    % daemon.returncode)
            try:
                self.query('dummy.nonexisting.zone')
                break
            except QueryRefused:
                break
            except DNS_SocketError as ex:
                # pydns >= 2.3.6
                wrapped_error = ex.args[0]
                if wrapped_error.errno != errno.ECONNREFUSED:
                    raise
            except DNS.DNSError as ex:
                # pydns < 2.3.6
                if str(ex) != 'no working nameservers found':
                    raise
            if retry > 10:
                raise DaemonError("rbldnsd does not seem to be responding")
            time.sleep(0.1)

    def _stop_daemon(self):
        daemon = self._daemon

        if daemon.poll() is None:
            daemon.terminate()
            retries = count()
            while daemon.poll() is None:
                retry = next(retries)
                if retry == 30:
                    daemon.kill()
                elif retry == 50:
                    raise DaemonError("can not kill stop rbldnsd")
                time.sleep(0.1)

        self._stdout.close()

        self._daemon = None
        if daemon.returncode != 0:
            raise DaemonError("rbldnsd exited with code %d"
                              % daemon.returncode)

    @property

    def no_ipv6(self):
        """ Was rbldnsd compiled with -DNO_IPv6?
        """
        # If rbldnsd was compiled with -DNO_IPv6, the (therefore
        # unsupported) '-6' command-line switch will not be described
        # in the help message
        cmd = [self.daemon_bin, '-h']
        proc = subprocess.Popen(cmd, stdout=subprocess.PIPE)
        help_message = proc.stdout.readlines()
        if proc.wait() != 0:
            raise subprocess.CalledProcessError(proc.returncode, cmd)

        return not any(line.lstrip().startswith(b'-6 ')
                       for line in help_message)


class TestRbldnsd(unittest.TestCase):
    def test(self):
        rbldnsd = Rbldnsd()
        test_zone = ZoneFile(lines=["1.2.3.4 :1: Success"])
        rbldnsd.add_dataset('ip4set', test_zone)
        with rbldnsd:
            self.assertEqual(rbldnsd.query('4.3.2.1.example.com'), 'Success')
            self.assertEqual(rbldnsd.query('5.3.2.1.example.com'), None)

if __name__ == '__main__':
    unittest.main()
